RenderPassEditor : Inclusion columns

murraystevenson opened the pull request:
This adds three new columns to the Render Pass Editor, providing control of the scene locations included in the render performed for each render pass.

https://github.com/GafferHQ/gaffer/assets/50844517/6c1092cd-ee41-4508-a56d-c7a4bfd6d9be

In keeping with the idea that Render Passes and the Render Pass Editor are a layer on top of base render functionality, this has been implemented as three new options configurable with set expressions defining "Inclusions", "Exclusions" and "Additional Lights".

- `render:inclusions` defines the scene locations included in the render.
- `render:additionalLights` provides the ability to separately control the lights contributing to the scene. These are combined with `render:inclusions` to produce the scene to render.
- `render:exclusions` are overrides, removing locations from the scene defined by `render:inclusions` and `render:additionalLights`.

These options are also configurable via the new "Render Set" section of a StandardOptions node, and can be used independently in scenes not making use of the Render Pass Editor.

This PR provides the base functionality for editing these options via columns in the Render Pass Editor, with further UI improvements to help manage these columns to come at a later date.

WangSnowchen commented:
Amazing! This feature is awesome, here's a table we created in our project to use for layering, to give you some reference.
![图片1](https://github.com/GafferHQ/gaffer/assets/62184149/5c467c8b-66ba-4581-a518-5d32ebc5cdef)
![图片2](https://github.com/GafferHQ/gaffer/assets/62184149/2529c567-583f-4d6e-81db-d2f1887f9b91)
However, we have found in practice in our projects that artists prefer to use this list style of layering.
![image](https://github.com/GafferHQ/gaffer/assets/62184149/b1627734-320a-4ffe-9fb9-847455e4ef54)

johnhaddon commented:
As part of testing this, I took a stab at updating the render pass example file to use the inclusions/inclusions rather than a StandardAttributes node : https://github.com/johnhaddon/gaffer/commit/14b1034c39afdee1548128dba015cad7d2548cad. Might be worth cherry-picking into this PR if it looks OK to you?

johnhaddon commented on python/GafferSceneUI/StandardOptionsUI.py:
```diff
@@ -77,16 +77,20 @@ def __cameraSummary( plug ) :
-		else :
-			return "None"
+		info.append( "Purposes {}".format( ", ".join( [ p.capitalize() for p in purposes ] ) if purposes else "None" ) )
+	if plug["inclusions"]["enabled"].getValue() :
+		info.append( "Inclusions {}".format( plug["inclusions"]["value"].getValue() ) )
+	if plug["exclusions"]["enabled"].getValue() :
+		info.append( "Exclusions {}".format( plug["exclusions"]["value"].getValue() ) )
+	if plug["additionalLights"]["enabled"].getValue() :
+		info.append( "Lights {}".format( plug["additionalLights"]["value"].getValue() ) )
 
-	return ""
+	return ", ".join( info )
```
Nitpick : Joining purposes with a comma and then joining that with other things with a comma is confusing :

![image](https://github.com/GafferHQ/gaffer/assets/1133871/8afd78ff-f2f5-4b1c-b99f-c363590c45b6)

Should we join purposes with a `/` instead? Or `&`?

johnhaddon reviewed (commented):
LGTM, minus the one nitpick and the suggestion to cherry-pick the docs update. Thanks Murray!

murraystevenson commented:
Thanks for the input! I've pushed an update with the following:

- Adjusted StandardOptions summary to join purposes with a `/`.
- Cherry picked your example update (thanks for that!) and added one small change to the annotation on the NameSwitch within the `RenderPassVisibility` box.
- Added an additional commit 62d7e03 to include light filters in `additionalLights` like we discussed offline.

Might be worth taking one final look, but this should be good to go.
